src/pip/_internal/models/link.py

```diff
@@ -170,12 +170,30 @@ def _ensure_quoted_url(url: str) -> str:
     and without double-quoting other characters.
     """
     # Split the URL into parts according to the general structure
-    # `scheme://netloc/path;parameters?query#fragment`.
-    result = urllib.parse.urlparse(url)
+    # `scheme://netloc/path?query#fragment`.
+    result = urllib.parse.urlsplit(url)
     # If the netloc is empty, then the URL refers to a local filesystem path.
     is_local_path = not result.netloc
     path = _clean_url_path(result.path, is_local_path=is_local_path)
-    return urllib.parse.urlunparse(result._replace(path=path))
+    return urllib.parse.urlunsplit(result._replace(path=path))
+
+
+def _absolute_link_url(base_url: str, file_url: str) -> str:
+    """Return an absolute file URL from a simple response.
+
+    If the file URL is already absolute, the function does nothing.
+    If the file URL is relative, it is joined with the base URL.
+    """
+    # If the file URL is already absolute, joining it with the page URL
+    # will do nothing. PyPI returns absolute URLs so we can avoid expensive
+    # urljoin() calls in the common case. (It would be technically more
+    # correct to parse the file URL and check if it has a scheme, but the
+    # slow URL parsing urljoin() does is what we're trying to avoid in the
+    # first place, so we only check for the http[s]:// prefix.)
+    if file_url.startswith(("https://", "http://")):
```

Comment:
Are we sure we don't need to support other schemes here? What about `file://`?

Reply:
This entire function is meant to be an optimization. We used to call `urljoin()` for every single file URL, but that was pointless for PyPI simple pages as their file URLs are already absolute (and thus `urljoin()` would return said URL unchanged). So it's not a matter of support, but rather is it worth to optimize for `file://`? How common are `file://` URLs in a simple index response? They'd only make sense with a local dev server (maybe devpi?)

I did just check and apparently relative `file://` URLs are not a thing, so it'd be totally fine to also check for that scheme, but I'm curious to whether you know where they'd show up.